File: day1/main.ml
let sample = {|
3   4
4   3
2   5
1   3
3   9
3   3
|}

let parse (input: string) = 
  String.trim input
  |> String.split_on_char '\n' 
  |> List.map (fun line -> 
    let parts = String.split_on_char ' ' line in
    List.filter (fun part -> not (String.equal (String.trim part) "")) parts
    |> List.map (fun part -> part |> String.trim |> int_of_string)
  )
  |> List.map (fun parts -> (List.nth parts 0, List.nth parts 1))
  |> List.split

let distance (lists: int list * int list) = 
  let (list1, list2) = lists in
  List.sort (-) list1
  |> List.combine (List.sort (-) list2)
  |> List.fold_left (fun acc (a, b) -> acc + abs (a - b)) 0


let count (query_list: int list) (el: int) = 
  query_list 
  |> List.find_all (fun x -> x == el) 
  |> List.length

let compare (lists: int list * int list) =
  let (list1, list2) = lists in
  List.fold_left (fun acc x -> acc + (x * (count list2 x))) 0 list1

let read_to_string filename =
  let ch = open_in_bin filename in
  let s = really_input_string ch (in_channel_length ch) in
  close_in ch;
  s

let () =
  parse sample |> distance |> Printf.printf "part 1 (sample): %i\n";
  parse sample |> compare |> Printf.printf "part 2 (sample): %i\n";
  read_to_string "input.txt" |> parse |> distance |> Printf.printf "part 1: %i\n";
  read_to_string "input.txt" |> parse |> compare |> Printf.printf "part 1: %i\n";

File: day1/main.nim
import std/strutils
import std/algorithm
import std/sequtils
import std/strformat

let sample = """
3   4
4   3
2   5
1   3
3   9
3   3
"""

proc parse(inputStr: string): (seq[int], seq[int]) =  
    var list1: seq[int]
    var list2: seq[int]
    for line in inputStr.strip.splitLines:
        let parts = line.split("   ")
        list1.add(parts[0].parseInt)
        list2.add(parts[1].parseInt)
    return (list1, list2)

proc distance(lists: (seq[int], seq[int])): int =
    let (list1, list2) = lists
    return list1.sorted.zip(list2.sorted).foldl(a + abs(b[0] - b[1]), 0)

proc compare(lists: (seq[int], seq[int])): int =
    let (list1, list2) = lists
    return list1.foldl(a + b * list2.countIt(it == b), 0)

echo &"part 1 (sample): {sample.parse.distance}"
echo &"part 1 (sample): {sample.parse.compare}"

let data = readFile("input.txt")
echo &"part 1: {data.parse.distance}"
echo &"part 1: {data.parse.compare}"
